### PR TITLE
refactor withPeerSelectionAction argument list

### PR DESCRIPTION
**Pull request by crocodile-dentist**
# Description

_description of the pull request, if it fixes a particular issue it should link the PR to a particular issue, see [ref](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)_

# Checklist

- Branch
    - [ ] Updated changelog files.
    - [ ] Commit sequence broadly makes sense
    - [ ] Commits have useful messages
    - [ ] The documentation has been properly updated
    - [ ] New tests are added if needed and existing tests are updated
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [ ] Reviewer requested

### CODE REVIEW

**Comment by crocodile-dentist** [comment]
issue #4745

**Comment by coot** [comment]
^ if there's an issue, it's enough to attach it in the `Development` section on the right, no need to add it to the project.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerSelectionActions.hs [hunk 1]
```diff
@@ -50,6 +53,25 @@ import Ouroboros.Network.PeerSharing (PeerSharingController, requestPeers)
...
+        :: Tracer m TraceLedgerPeers
+    }
+
+data PeerSelectionActionExtraArgs m =
+  PeerSelectionActionExtraArgs {
+      daReadLocalRootPeers  :: STM m [(HotValency, WarmValency, Map RelayAccessPoint (PeerAdvertise, PeerTrustable))]
+    , daReadPublicRootPeers :: STM m (Map RelayAccessPoint PeerAdvertise)
+    , daReadUseBootstrapPeers :: STM m UseBootstrapPeers
+    , daOwnPeerSharing      :: PeerSharing
+    }
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerSelectionActions.hs:74]: Could you also add some haddocks?

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerSelectionActions.hs [hunk 2]
```diff
@@ -3,6 +3,7 @@
 {-# LANGUAGE LambdaCase          #-}
 {-# LANGUAGE NamedFieldPuns      #-}
 {-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE RecordWildCards     #-}
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerSelectionActions.hs:6]: Please avoid `RecordWildCards`: it often hides where arguments are coming from and going to, which makes it more difficult to read the code - we try to optimise for reading, not writting.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerSelectionActions.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (crocodile-dentist) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerSelectionActions.hs]: Good to know.

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs [hunk 3]
```diff
@@ -980,17 +980,12 @@ runM Interfaces
             -> m c
           withPeerSelectionActions' =
               withPeerSelectionActions
-                  dtTraceLocalRootPeersTracer
-                  dtTracePublicRootPeersTracer
-                  dtTraceLedgerPeersTracer
+                  PeerSelectionActionTracers {..}
```
[the remark this answers is not in the document]
Comment (crocodile-dentist) [reply ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs]: Good to know.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerSelectionActions.hs [hunk 4]
```diff
@@ -98,17 +110,12 @@ withPeerSelectionActions
...
   -> m a
 withPeerSelectionActions
-  localRootTracer
-  publicRootTracer
-  ledgerPeersTracer
+  PeerSelectionActionTracers {..}
   toPeerAddr
   dnsActions
   readPeerSelectionTargets
   readLedgerStateJudgement
```
[the remark this answers is not in the document]
Comment (crocodile-dentist) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerSelectionActions.hs]: I've repackaged all arguments into three logical records, minus the continuation ofc.

File: ouroboros-network/CHANGELOG.md [hunk 5]
```diff
@@ -54,6 +54,8 @@
 
 * Disable mean reward for new peers
 
+* Refactored withPeerSelectionActions argument list
+
```
[the remark this answers is not in the document]
Comment (crocodile-dentist) [reply ouroboros-network/CHANGELOG.md]: Correct, I've moved the change into breaking section.

**Review by coot — CHANGES_REQUESTED** [review]
Very nice, a few minor comments.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerSelectionActions.hs [hunk 6]
```diff
@@ -41,15 +42,42 @@ import Ouroboros.Network.PeerSelection.PeerSharing (PeerSharing)
...
 
+-- | Record of parameters for withPeerSelectionActions independent of diffusion mode
+--
+data PeerSelectionActionsArgs peeraddr peerconn exception m = PeerSelectionActionsArgs {
+  psLocalRootPeersTracer      :: Tracer m (TraceLocalRootPeers peeraddr exception),
+  psPublicRootPeersTracer     :: Tracer m TracePublicRootPeers,
+  psReadTargets               :: STM m PeerSelectionTargets,
+  -- ^ peer selection governor know, established and active targets
+  psJudgement                 :: STM m LedgerStateJudgement,
+  -- ^ are we there yet?
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerSelectionActions.hs:60]: :smile:
```suggestion
  -- ^ Is consensus close to current slot?
```

**Review by coot — APPROVED** [review]
LGTM, thanks @crocodile-dentist.

**Pull request merged**